File: backend/workflow/api/src/main/java/br/com/systec/taskflow/workflow/api/service/WorkflowService.java
package br.com.systec.taskflow.workflow.api.service;

import br.com.systec.taskflow.commons.exceptions.BaseException;
import br.com.systec.taskflow.workflow.api.filter.WorkflowFilter;
import br.com.systec.taskflow.workflow.api.vo.StatusVO;
import br.com.systec.taskflow.workflow.api.vo.WorkflowVO;
import org.springframework.data.domain.Page;

public interface WorkflowService {

    WorkflowVO create(WorkflowVO workflow) throws BaseException;

    WorkflowVO update(WorkflowVO workflowVO) throws BaseException;

    WorkflowVO findById(Long id) throws BaseException;

    Page<WorkflowVO> findByFilter(WorkflowFilter filter) throws BaseException;

    StatusVO findStatusById(Long statusId) throws BaseException;

    StatusVO findInitialStatusByWorkflowId(Long workflowId) throws BaseException;
}

File: frontend/src/administrator/workflow/service/workflowService.ts
import Page from "@/@core/utils/page";
import { http } from "@/plugins/axios/axios";
import Workflow from "../model/workflow";
import WorkflowResponse from "../model/workflowResponse";

export default class WorkflowService {
    private _endpoint: string;

    constructor() {
        this._endpoint = "/api/v1/workflows";
    }

    async save(workflow: Workflow): Promise<Workflow> {
        if (workflow.id === null) {
            return this.create(workflow);
        }

        return this.update(workflow);
    }

    private async create(workflow: Workflow): Promise<Workflow> {
        const response = await http.post(this._endpoint, workflow);

        return response.data;
    }

    private async update(workflow: Workflow): Promise<Workflow> {
        const response = await http.put(this._endpoint, workflow);

        return response.data;
    }

    async findById(workflowId: number): Promise<Workflow> {
        const response = await http.get(`${this._endpoint}/${workflowId}`);

        return response.data;
    }

    async findByFilter(): Promise<Page<WorkflowResponse>> {
        const response = await http.get(`${this._endpoint}/filter`);

        return response.data;
    }
}
